### PR TITLE
gh-112510: Add `readline.backend` for the backend readline uses

### PR DESCRIPTION
**Pull request by gaogaotiantian**
* Issue: gh-112510

----
:books: Documentation preview :books:: https://cpython-previews--112511.org.readthedocs.build/

### CODE REVIEW

**Comment by gaogaotiantian** [comment]
@encukou @corona10 could you take a look? Thanks!

File: Modules/readline.c [hunk 1]
```diff
@@ -1568,6 +1571,11 @@ PyInit_readline(void)
         goto error;
     }
 
+    if (PyModule_AddStringConstant(m, "backend", backend) < 0)
+    {
+        goto error;
+    }
```
Comment (corona10) [inline Modules/readline.c]:
```suggestion
    if (PyModule_AddStringConstant(m, "backend", backend) < 0) {
        goto error;
    }
```

pep7

File: Doc/library/readline.rst [hunk 2]
```diff
@@ -27,23 +27,26 @@ Readline library in general.
...
-  If you use *editline*/``libedit`` readline emulation on macOS, the
+  If you use ``editline``/``libedit`` readline emulation on macOS, the
   initialization file located in your home directory is named
   ``.editrc``. For example, the following content in ``~/.editrc`` will
   turn ON *vi* keybindings and TAB completion::
 
     python:bind -v
     python:bind ^I rl_complete
 
+.. data:: backend
```
Comment (corona10) [inline Doc/library/readline.rst:46]: Out of curiosity, do we have to export the backend name as the public?

File: Lib/test/test_pdb.py [hunk 3]
```diff
@@ -3293,7 +3293,7 @@ def setUpClass():
         # Ensure that the readline module is loaded
         # If this fails, the test is skipped because SkipTest will be raised
         readline = import_module('readline')
-        if readline.__doc__ and "libedit" in readline.__doc__:
+        if readline.backend == "editline":
```
Comment (corona10) [inline Lib/test/test_pdb.py:3296]: For just internal usage, why not just use readline._backend?

File: Modules/readline.c (hunk 1 above)
Comment (gaogaotiantian) [reply Modules/readline.c]: Ah, I was following the line above it. Sure that makes sense.

File: Doc/library/readline.rst (hunk 2 above)
Comment (gaogaotiantian) [reply Doc/library/readline.rst:46]: Because users will have the same issue - we have a big note section to explain this in our docs.

Comment (corona10) [reply Doc/library/readline.rst:46]: Ah okay @encukou commented about this: https://github.com/python/cpython/pull/107748#discussion_r1395432432

**Review by corona10 — APPROVED** [review]
LGTM from my view. But I will leave it to @encukou since we have to export `backend` as the public. Once we export it we should maintain it for a long time and Petr is the expert in this area.

**Comment by encukou** [comment]
Thank you!
My main worry here is that we might get into a situation like `sys.version`/`sys.version_info`, where the earlier “stringly typed” attribute has a better name.
Are we likely to need something like `backend.name` & `backend.version` in the future? If we do, will it be OK to keep `backend` and a new `backend_version`?
Or am I overthinking this?

---

Could you add a test to assert `backend` is either `"readline"` or `"editline"`?

**Comment by gaogaotiantian** [comment]
> Thank you! My main worry here is that we might get into a situation like `sys.version`/`sys.version_info`, where the earlier “stringly typed” attribute has a better name. Are we likely to need something like `backend.name` & `backend.version` in the future? If we do, will it be OK to keep `backend` and a new `backend_version`? Or am I overthinking this?
> 
> Could you add a test to assert `backend` is either `"readline"` or `"editline"`?

Very helpful insight! From my observation, the difference between different versions of the same library is significantly less important than the difference between the two libraries. I can't guarantee that the version of the backend will not ever be used at all in the future, but I would guess that's a very rare case. Both `readline` and `editline` seems to be relatively consistent between versions.

That being said, I think using a single string type for `readline.backend` is reasonable, and it probably makes the life easier for the users compared to having a `(name, version)` tuple for the attribute and the users would have to do `readline.backend[0] == 'readline'` in most of the cases.

If in the future, `readline` or `editline` has a breaking change in some version that affects many users, having a `readline.backend_version` is not the end of the world.

**Comment by gaogaotiantian** [comment]
I think we are done with the review for this PR? We should merge this before #107748 so it can use the `backend` instead of checking `__doc__`

**Review by encukou — DISMISSED** [review]

File: Doc/library/readline.rst [hunk 4]
```diff
@@ -27,23 +27,26 @@ Readline library in general.
...
+
+   The name of the underlying Readline library being used, either
+   ``"readline"`` or ``"editline"``.
```
Comment (encukou) [inline Doc/library/readline.rst:49]:
```suggestion
   ``"readline"`` or ``"editline"``.
   
   .. versionadded:: 3.13
```

**Comment by encukou** [comment]
One more thing: the docs entry needs a `versionadded`. I updated the PR directly to avoid a round of review ping-pong; hope you don't mind.

(And I fell into the trap of thinking this is simple enough to use GitHub UI rather than a local clone...)

**Comment by gaogaotiantian** [comment]
Thanks for the quick review!

**Pull request merged**